### ISSUE TITLE
NaN param values introduced for multi gpu training between 3a52bdf and e8e1c6a

### ISSUE THREAD

**Issue by skaae** [microsoft/dion#76]
Hi, 

First thank you for great library!

After updating from commit 3a52bdf  to e8e1c6a we started having training jobs fail where the parameters will be NaN. I have indentified that the gradients are ok, but parameters will be Nan after optimizer.step has been called. We train using 6 blackwell 6000 gpus using DDP and nccl. 

We have 6 blocks in our network. the failing layer is always the linear layer with shape [512, 1408] in one of the blocks. Its not deterministic what block that fails, but its always the [512, 1408]  that fails. Its also not deterministic when the error happens, but usually after 0-30 min of training. 

I have so far confirmed that the gradients are fine until `optimizer.step()` is called. After the optimizer.step() the layer weight with shape [512, 1408]  will be all Nans. I therefore suspect its related to the update the of the Newton-Shultz package or maybe Quack Kernels being bumped? Alternatively the megabatch PR. 

Sorry for the not super specific bug report.
I'm happy to dump some better debug data, but I would need some guidance on what to dump. 

```
Driver Version: 590.48.01      CUDA Version: 13.1  
torch                              2.11.0+cu130


OK version:
dion commit 3a52bdf
quack kernels pinned to 0.3.10

Failing version:
dion commit e8e1c6a
```

**Comment by alint77** [comment]
Have you tested PR #75 ?

**Comment by JohnLangford** [comment]
This is super-important to nail, but I need some help because I can't reproduce.
--Have you tried a bisection search for the exact commit?
--what are the flags you are using exactly?
--Have you tried swapping the optimizer cores (gram ns with/without triton vs. polar express)?

**Comment by skaae** [comment]
I Spent yesterday narrowing it down to the optimizer step.  



**--Have you tried swapping the optimizer cores (gram ns with/without triton vs. polar express)?**
No. Do you have some combinations you want me to try? 

I can also just dump the input/ouput updates of the normuon internals? I'm not super familiar with NorMuon but if you can point me to the most likely place where this could happen inside NorMuon, then I can check for nans and save the input that resulted in Nans. 

From looking at the code I was about to check for nans around this function: 

```
    U = yield from megabatch_orthogonalize_async(
        U,
        comm_dim=comm_dim,
        device_rank=device_rank,
        world_size=world_size,
        process_group=process_group,
        newton_schulz_func=newton_schulz_func,
        flatten=flatten,
        epsilon=epsilon,
    )
```

And save the input if I detect any nans? 

**--Have you tried a bisection search for the exact commit?**
I should be able to narrow it down to specific commit. I can do that or try to get some faulty input, what ever you think is most likely to identify the issue? 


**--what are the flags you are using exactly?**
I assume its for the optimizer? 

```
optimizer = NorMuon(
            param_groups,
            distributed_mesh=process_group,  # world- 6 gpu's
            weight_decay=weight_decay,  # 0.1
            nesterov=True,
            # adjust_lr="spectral_norm",
            use_gram_newton_schulz=True,
            use_triton=True,
            use_polar_express=True,
        )
```

**Comment by JohnLangford** [comment]
Some things to try: 
use_polar_express=False
use_triton=False
use_gram_newton_schulz=False
Pin just quack to 0.3.10 on e8e1c6a and patch gns locally to pass alpha=1.0, beta=1.0 
Bisect over the 8 commit interval to find the offender.  
Go ahead with the nan-guard to capture an offending input.

Basically: just try a bunch of variations so we can isolate the offending code.

**Comment by NoahAmsel** [comment]
> Go ahead with the nan-guard to capture an offending input.

It's possibly a numerical blow up due to gram newton schulz. If so, having an offending input would be the quickest way to check and to remediate

**Comment by skaae** [comment]
ok. does the nan guard also work for multi gpu setting?

**Comment by NoahAmsel** [comment]
You might want to save the offending input to a file whose name depends on the GPU device number. that way multiple GPUs aren't competing to write their faulty inputs to the same file

**Comment by skaae** [comment]
ok, figured I had to do something like that. I'll see if i can narrow it down to a single commit and capture som data that triggers the error.

**Comment by skaae** [comment]
It looks like 9a78633cb6428632b7536decbe0da2cc995cc88d is the latest commit that is ok. It has run more than 5 epochs now and the two failing commits ran less than one epoch. 



FAILS: e8e1c6ae95abe61bdd687566cfeaa9a2168cd49e  
FAILS: 9f7897d413432c44e527722b4f39b59a4dc46e04
OK:     9a78633cb6428632b7536decbe0da2cc995cc88d

**Comment by alint77** [comment]
> It looks like 9a78633cb6428632b7536decbe0da2cc995cc88d is the latest commit that is ok. It has run more than 5 epochs now and the two failing commits ran less than one epoch. 
> 
> 
> 
> FAILS: e8e1c6ae95abe61bdd687566cfeaa9a2168cd49e  
> FAILS: 9f7897d413432c44e527722b4f39b59a4dc46e04
> OK:     9a78633cb6428632b7536decbe0da2cc995cc88d
> 
> 
> 
> 

So does that mean with gram_ns = false, it won't get NaN on latest commit? Since the eraliest offending commit is related to gram_ns

**Comment by skaae** [comment]
I will try 9f7897d413432c44e527722b4f39b59a4dc46e04 with `use_gram_newton_schulz=False`. 
Im in Europe so i might report back tomorrow.

**Comment by NoahAmsel** [comment]
Ah so it's something about quack 0.4.1. When we have an example input, I can start debugging it on that end. Agreed that gram_ns = false should solve the problem

**Comment by JohnLangford** [comment]
@skaae any updates?

**Comment by skaae** [comment]
Hi, 
I spent today trying to capture some input using https://github.com/microsoft/dion/commit/e8e1c6ae95abe61bdd687566cfeaa9a2168cd49e. But I cant trigger the error anymore, which I don't fully understand. I don't think i changed anything and I'm using the exact same hardware. 

I think I will start using HEAD and reopen this if I manage to capture some data?
Unless you want me to try something?

**Comment by JohnLangford** [comment]
I attempted to reproduce on our B200s and have failed.  This seems likely something quite specific.

**Comment by skaae** [comment]
Yes :) but it happend again on head (4dc5728fb7e605b658ac44e0720ee87fe25b806f). 
I probably have time to try to capture some debugging date sometime next week. until then i will try to disable ns and see if that helps.

**Comment by JohnLangford** [comment]
Are PR 78 & 79 useful here?

**Comment by skaae** [comment]
Hi, 
Sorry for the silence. We have discovered a hardware error on the machine related to to the motherboard. I'm Not really sure what the hardware error was as it didn't show up in dmesg or as an nccl error. The machine would refuse to boot if we added moret than 6 gpu's. I swapped the motherboard and have not had any issues with dion since. So i assume its some sort of nccl error that triggered the nans?  

Thank you for looking into this.

**Comment by JohnLangford** [comment]
Ok, thanks for letting us know.

**Comment by skaae** [comment]
I have been running for ~1 week using `use_gram_newton_schulz=False` without any issues. We tried switching to use_gram_newton_schulz=True and the NaN issue started to appear again. I.e it looks like for our particular use-case theres something with use_gram_newton_schulz=True that causes NaN issues.

**Comment by NoahAmsel** [comment]
Thanks again for letting us know. Are you able to capture the offending inputs this time?

**Comment by donglixp** [comment]
https://github.com/microsoft/dion/pull/102
